Implement mount option support

laenion opened the pull request:
Implements #872, using the _mkfs_ option implementation as a reference.

With this new `mountOptions` entry it is possible to add mount options for each file system, used during the `mount` stage.

ajeddeloh commented:
ok to test

lucab commented on internal/exec/stages/mount/mount.go:
```diff
@@ -117,10 +118,19 @@ func (s stage) mountFs(fs types.Filesystem) error {
 	return nil
 }
+
+func translateOptionSliceToString(opts []types.MountOption) string {
```
Minor: it would be useful to take `,` as an argument here, and also have some basic tests to cover the empty-slice case and a simple populated-slice too.

laenion replied:
Will add the separator as a separate argument.

Regarding the slice: I'm no Go expert, but from what I know and see Go's `Join` function will handle both the empty and single slice automatically (i.e. return an empty string / a string without comma). Or do you mean testing the functionality via Blackbox tests?

lucab commented:
PR and logic seems overall fine to me.
I have the fear that this will confuse a lot people, who will start assuming that they can put here options that will be in effect in the rootfs and for any subsequent boot.
How can we make obvious that these are just ephemeral options for Ignition and initramfs and first-boot only?

laenion commented:
Thanks for the reviews and sorry for the late reply, I've been on vacation last week.

> I have the fear that this will confuse a lot people, who will start assuming that they can put here options that will be in effect in the rootfs and for any subsequent boot.
> How can we make obvious that these are just ephemeral options for Ignition and initramfs and first-boot only?

The most obvious solution would probably be to mention that to the documentation explicitly. Maybe by adding a sentence such as
> Please note that these options will only be applied during Ignition's `mount` stage (i.e. using this option will not result in automatic creation of corresponding mount units or fstab entries).

to the documentation of the `mountOptions` option?

I'll fix the remaining issues shortly.

laenion commented:
Added the translator and added the option to the IgnoreDuplicates list. Also added the separator as a separate argument to `translateOptionSliceToString`.

ajeddeloh commented on internal/exec/stages/mount/mount.go:
```diff
@@ -117,10 +118,19 @@ func (s stage) mountFs(fs types.Filesystem) error {
-	if err := s.Logger.LogOp(func() error { return syscall.Mount(fs.Device, path, *fs.Format, 0, "") },
-		"mounting %q at %q with type %q", fs.Device, path, *fs.Format,
+	args := translateOptionSliceToString(fs.MountOptions, ",")
+	if err := s.Logger.LogOp(func() error { return syscall.Mount(fs.Device, path, *fs.Format, 0, args) },
```
Sorry for the late review; I was at a conference then some regressions prevented me from building an image to test for a bit. Looks like this only works with FS specific options, but not ones that end up as flags for the fourth argument. For example, `ikeep` is xfs specific and works, but `noexec` is not and doesn't since that ends up as the flag `MS_NOEXEC`. So this needs to learn how to handle those separately. It would be good to add a blackbox test for both types of args as well.

ajeddeloh replied:
https://github.com/karelzak/util-linux/blob/master/libmount/src/optstr.c#L651 and that file in general looks useful, though it would need cgo.

laenion replied:
Just that it's not part of Go, but an additional dependency. I wasn't aware that `syscall` is deprecated, so this is indeed a non-issue.

ajeddeloh replied:
It'd be good to not vendor something not really meant as a library especially something large like docker, so I think I'd prefer exec'ing mount if you're opposed to the cgo implementation. @lucab do you have strong opinions here?

laenion replied:
I'm not opposed to use _libmount_, it's just that I usually try to find a native library before implementing something myself.

FYI: The size of the Ignition executable is increasing by ~100 KB when using the Docker module.

ajeddeloh replied:
It's less about the actual binary size and more about needing to keep up to date with another dependency as well as the go module tooling being bad and wanting to pull in more than necessary to `vendor/`.

lucab replied:
@ajeddeloh either directions are fine. As we are already shelling out to `mount` in other places, I think that would be path of least resistance here too.

laenion replied:
Implemented as a shell command as suggested.

laenion commented:
Added a positive and a negative Blackbox test case for the mount stage.

ajeddeloh commented on tests/negative/filesystems/mount_filesystem.go:
```diff
@@ -0,0 +1,61 @@
+// Copyright 2017 CoreOS, Inc.
```
2019 Red Hat

ajeddeloh commented on tests/positive/filesystems/mount_filesystem.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2017 CoreOS, Inc.
```
2019 Red Hat
